runtime/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,161 @@
+use {
+    crate::bank::Bank,
+    solana_sdk::slot_history::Slot,
+    std::{
+        sync::{Arc, Condvar, Mutex, Weak},
+        time::{Duration, Instant},
+    },
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader bank initiates (=PoH-initiated)
+///     2. A leader slot completes (=PoH-completed)
+#[derive(Debug, Default)]
+pub struct LeaderBankNotifier {
+    /// Current state (slot, bank, and status) of the system
+    state: Mutex<SlotAndBankWithStatus>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// The leader bank is not currently available. Either not initialized, or PoH-completed bank.
+    #[default]
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::InProgress) {
+            return;
+        }
+
+        *state = SlotAndBankWithStatus {
+            status: Status::InProgress,
+            slot: bank.slot(),
+            bank: Arc::downgrade(bank),
+        };
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if
+    ///     1. the status was not already `StandBy` and
+    ///     2. the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
```

Comment:
also, `slot < state.slot` could turn out to be problematic. i think there's a possibility where leader_slot+1 is created later leader_slot+3 for example, in th event of excessive forking and poh-reset.

considering, set_in_progress is single threaded by definition with replay stage thread and poh thread, i'd rather add `assert_eq!(self.state.slot, slot);` just before the following `if` for strictest sanity check (i.e. before idempotent bail-out). i know, i know, panic are bad. but imo, better than debugging a race condition on mainnet-beta, I'd say.

Reply:
done: 52632ee4d579872be1957c45086f3825af561c64